### PR TITLE
Feature/api security

### PR DESCRIPTION
**Pull request by danewrye**
- Added loci_repeats parameter to api
- Added custom exception handling for malformatted loci_repeats
- Updated API Readme to fix previous inconsistencies and add information about new parameters
- Added API Rate Limiting based on remote address (see library used at https://flask-limiter.readthedocs.io/en/stable/)

### CODE REVIEW

File: api/API.md [hunk 1]
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
...
 **Input Parameters (URL):**
 
 -   (optional) `iterations` (integer): The number of iterations to run the algorithm for on the data; higher values are more accurate but take longer to complete. Defaults to 10000.
 
 **Input Parameters (Data):**
 
--   `file` (.csv file, FormData): The file containing the testing data to be processed.
+-   `file` (.xlsx file, FormData): The file containing the testing data to be processed.
+
+**Input Parameters (URL):**
```
Comment (JHDeerin) [inline api/API.md]: Don't create a 2nd "Input Parameters (URL)" section; instead, add this as a bullet point under the 1st "Input Parameters (URL)"

File: api/API.md [hunk 2]
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
...
+
+-   (optional) `locirepeat` (list of integers): List of integers representing loci repeats. Each loci repeat must be individually entered for the request. Please ensure the correct amount of values are entered.
```
Comment (JHDeerin) [inline api/API.md]: Locirepeats shouldn't be optional; it should be a required parameter (file can't be processed correctly without it)

File: api/API.md [hunk 3]
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
...
 
 **Example Usage:**
 
 In cURL:
 
 ```bash
-curl -F "file=@example.csv" http://localhost:5000/api/v1/recrudescences?iterations=10
+curl -F "file=@example.xlsx" "http://localhost:5000/api/v1/recrudescences?iterations=10&locirepeat=2&locirepeat=2&locirepeat=3"
 ```
+If you are using curl and are entering multiple query parameters, ensure you use quotes around the request
```
Comment (JHDeerin) [inline api/API.md]: This should already be known by people who are using curl and can easily be looked up, so I don't think we need to include it; if you still want to, I would reformat it slightly to be in italics/etc. (to make it clear it's a side-note)

File: api/API.md [hunk 4]
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
...
-Given an input `.csv` file with [drug testing data in the proper format (TODO)](), return estimates of which patients specified in the file are reinfections or recrudescences.
+Given an input `.xlsx` file with [drug testing data in the proper format (TODO)](), return estimates of which patients specified in the file are reinfections or recrudescences.
 
-*Method(s):* `PUSH`
+*Method(s):* `POST`
 
 *URL Endpoint:* `/api/v1/recrudescences`
 
-#### PUSH
+#### POST
```
Comment (JHDeerin) [inline api/API.md:49]: Thank you! Good catch!

File: api/api.py [hunk 5]
```diff
@@ -44,13 +50,22 @@ def too_large(e):
...
+     ]
     def post(self):
         '''
-        Takes in a posted CSV data file, analyzes it to determine if each
+        Takes in a posted xlsx data file, analyzes it to determine if each
         case is a recrudescence/reinfection, and returns the calculated results
         '''
         uploaded_file = request.files.get('file')
-        iterations = request.args.get('iterations', default=10000, type=int)
+        iterations = request.args.get('iterations', default=50, type=int)
```
Comment (JHDeerin) [inline api/api.py]: Change the default back to 10000 (this should be set on the frontend, but might as well do it for future API use)

File: api/api.py [hunk 6]
```diff
@@ -44,13 +50,22 @@ def too_large(e):
...
+        loci_repeats = request.args.getlist('locirepeat', type=int)
+        if not loci_repeats:
+            loci_repeats = [2,2,3,3,3,3,3]
```
Comment (JHDeerin) [inline api/api.py]: Remove this; this `loci_repeats` value is specific to the Angola example Mat gave us, so I would make loci_repeats required and throw an error if it isn't included instead

File: api/algorithm_site_instance.py [hunk 7]
```diff
@@ -143,6 +145,9 @@ def __init__(
         self.ids = recrudescence_utils.get_sample_ids(genotypedata_RR, 'Day 0')
         self.locinames = pd.unique(genotypedata_RR.columns[1:].str.split("_").str[0])
 
+        if len(self.locinames) > len(locirepeats):
+            raise ValueError("Loci repeats has an insufficient number of entries")
```
Comment (JHDeerin) [inline api/algorithm_site_instance.py]: Could you make a custom exception for improper loci repeats in this file, and then use that? I know that's annoying, but ValueErrors can be thrown by other things that might lead to confusing error messages about loci repeats that have nothing to do with the loci

File: api/api.py [hunk 8]
```diff
@@ -1,11 +1,14 @@
 from collections import OrderedDict
+from typing import List
 import datetime
 import math
 import os
 import time
 
 from flask import Flask, request
 from flask_restful import Resource, Api
+from flask_limiter import Limiter
```
Comment (JHDeerin) [inline api/api.py:10]: Is `flask_limiter` a new library that has to be installed via `pip`? If so, can you add that library to `API.md` at the top, under the install instructions?

File: api/api.py [hunk 9]
```diff
@@ -44,13 +50,22 @@ def too_large(e):
 
 
 class RecrudescenceTest(Resource):
+
+    decorators = [
```
Comment (JHDeerin) [inline api/api.py:55]: Where are these decorators actually applied? Quickly checking, this list syntax doesn't seem to be one of the standard ways of applying limits to a flask function (https://flask-limiter.readthedocs.io/en/stable/)

Have you tested to make sure the limiter is actually being used right now?

**Review by JHDeerin — CHANGES_REQUESTED** [review]
Overall this looks good! There's a few tweaks to the documentation I think are needed and making locirepeats a required parameter seems necessary, and I want to double-check these rate limits actually are working, but the core seems solid.

File: api/API.md (hunk 2 above)
Comment (JHDeerin) [reply api/API.md]: Also, here's a description from `define_alleles` about what locirepeats actually are that might be useful to include: `a vector of length <number of loci> with the type of each locus (dinucleotide, trinucleotide, etc. repeats)`

File: api/API.md (hunk 1 above)
Comment (danewrye) [reply api/API.md]: Makes sense, change made

File: api/API.md (hunk 2 above)
Comment (danewrye) [reply api/API.md]: Done, and thanks for the description! It is much better than the one I had in there

File: api/API.md (hunk 3 above)
Comment (danewrye) [reply api/API.md]: True. I will delete it.

File: api/api.py (hunk 5 above)
Comment (danewrye) [reply api/api.py]: Done. Of all things to forget to do...

File: api/api.py (hunk 6 above)
Comment (danewrye) [reply api/api.py]: Done. Good improvement over the specific default

File: api/api.py (hunk 8 above)
Comment (danewrye) [reply api/api.py:10]: Good catch, completely forgot to add

File: api/api.py (hunk 9 above)
Comment (danewrye) [reply api/api.py:55]: Yes, it has been tested and works properly. If you also want to test it, just turn the max request limit down very low. There were problems with the standard ways outlined in the docs- if you want I can try to find the stackoverflow post i referenced in implementing that line

File: api/algorithm_site_instance.py (hunk 7 above)
Comment (danewrye) [reply api/algorithm_site_instance.py]: Change made, and no problem! Completely see the point about clarity, and the effort to implement was minimal. With this change, I think it should be ready for re-review

File: api/API.md (hunk 4 above)
Comment (danewrye) [reply api/API.md:49]: No problem!

File: api/api.py [hunk 10]
```diff
@@ -109,7 +110,7 @@ def _get_test_results_json(
 
         try:
             results = test_run.run_algorithm(iterations, burnin, record_interval)
-        except ValueError:
+        except LociRepeatError:
             return error_response('Loci repeats has an insufficient number of entries', 400)
```
Comment (JHDeerin) [inline api/api.py]: Testing this, it doesn't appear to actually be caught/printed in the response; I just get a generic "A problem occurred while the server was processing this data." Also, this method is currently expected to just return the JSON by `post`, not the HTML status code as well (despite what the docstring comment says).

Could you hunt down what's causing that (possibly the error doesn't get caught and a different error is thrown in AlgorithmInstance/etc?), and update the `post` method to use the HTML code from this as well?

**Review by JHDeerin — CHANGES_REQUESTED** [review]
Everything looks good except for the failing error message; fix that bug, and that's it!

File: api/api.py (hunk 10 above)
Comment (danewrye) [reply api/api.py]: I fixed the error, but I'm not 100% sure what you mean with the HTML status codes. I removed the only line in _get_test_results_json that returned an html status code (un-necessary, could be handled in post). Is that what you were asking? I don't think there are any HTML status codes from this method that can be returned. Or did you just want me to edit the docstring to reflect this?

**Review by JHDeerin — APPROVED** [review]
Great! Let's merge this in!

**Pull request merged**